update.sh: further speed up `brew update`.

MikeMcQuaid opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Tweak the logic further to make the no-op case even faster.

Before:
```
brew update  1.10s user 1.05s system 92% cpu 2.325 total
brew update --preinstall  0.60s user 0.77s system 96% cpu 1.433 total
```

After:
```
brew update  0.60s user 0.34s system 83% cpu 1.132 total
brew update --preinstall  0.18s user 0.20s system 120% cpu 0.318 total
```

These times are now fast enough to avoid any further special-casing for `--preinstall`, roll it out to users by default and not print a message unless we've actually found some updates.

UniqMartin commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -448,7 +426,38 @@ EOS
+      continue
+    fi
+
+    TAP_VAR=$(repo_var "$DIR")
```
I missed this in my style PR. Could you please quote this?

UniqMartin commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -448,7 +426,38 @@ EOS
+      INITIAL_REVISION="$(git rev-parse -q --verify "$UPSTREAM_BRANCH")"
+      CURRENT_REVISION="$(read_current_revision)"
+      export HOMEBREW_UPDATE_BEFORE"$TAP_VAR"="$INITIAL_REVISION"
+      export HOMEBREW_UPDATE_AFTER"$TAP_VAR"="$CURRENT_REVISION"
```
`$TAP_VAR` is being used here before it is populated (further below). However, I wonder if we should just remove the `--simulate-from-current-branch` option and the associated code from the update. It complicates things and makes the code a bit harder to understand and is probably rarely used, if ever.

If we wanted to retain the functionality (use local changes to test the migration code that kicks in after an update), I think this could nowadays better live in a `dev-cmd` that is custom-tailored to this use case. It could refrain from fetching and doing other modifications to the repositories and instead just set up the necessary environment variables and then invoke `brew update-report` (which will do most of the work).

UniqMartin commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -387,10 +362,15 @@ EOS
-        # Skip taps without formulae.
-        FORMULAE="$(find "$DIR" -maxdepth 1 \( -name "*.rb" -or -name Formula -or -name HomebrewFormula \) -print -quit)"
-        [[ -z "$FORMULAE" ]] && exit
+        [[ -n "$(find "$DIR/.git/FETCH_HEAD" -type f -mtime -60s)" ]] && exit
```
I guess this needs a `2>/dev/null` per #671 and to resolve the merge conflict.

UniqMartin commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -448,7 +426,38 @@ EOS
+    fi
+
+    TAP_VAR=$(repo_var "$DIR")
+    if [[ -n "$(find "$DIR/.git/FETCH_HEAD" -type f -mtime -60s)" ]]
```
Is this a good indicator that something was fetched? Would comparing the `origin/master` commit hash before and after the fetch maybe be a better indicator? (Not sure what the best solution is, but this feels like a strange criterion. It also impacts the non-`--preinstall` case which it probably shouldn't?)

However, if we stick with this, this should get a `2>/dev/null` just in case.

UniqMartin commented:
I think this change is basically sound, but it is also pretty invasive. Can you elaborate a little bit on what part of it is responsible for the massive speedup and why?

MikeMcQuaid commented:
I've addressed your commits and made it faster still for the `--preinstall` case as a result.

> Can you elaborate a little bit on what part of it is responsible for the massive speedup and why?

Most of it 😀. In short like most dramatic speed-up: do less where possible. Partially avoid recalculating the same values in multiple places but the big speedup: don't try to `status`/`stash`/`merge`/`rebase` when we know nothing has changed (i.e. calling the `merge_or_rebase` method [formerly `pull`]). These are all (relatively) expensive no-ops as they operate on the filesystem data rather than just with variables and filesystem metadata. We've already removed the unnecessary expensive `git fetch` and now `curl` operations too.

MikeMcQuaid commented:
CC @ilovezfs who will probably be happy with this speedup and was encouraging me to 🚢 this even before it was this fast.

ilovezfs commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -378,19 +378,34 @@ EOS
-        # Skip taps without formulae.
-        FORMULAE="$(find "$DIR" -maxdepth 1 \( -name "*.rb" -or -name Formula -or -name HomebrewFormula \) -print -quit)"
-        [[ -z "$FORMULAE" ]] && exit
+        [[ -n "$(find "$DIR/.git/FETCH_HEAD" -type f -mtime -60s 2>/dev/null)" ]] && exit
```
should the hard-coded 60 be a configurable number instead?

MikeMcQuaid replied:
Interesting idea. Using an environment variable?

ilovezfs replied:
That seems simplest until there's some more formal config file and/or mechanism for brew generally like a `brew get foo` and `brew set foo=bar` or however that ends up working. Alternatively, it could be exposed now with a specific command like `brew analytics` is for the on/off value.

ilovezfs replied:
Perfect, thank you!

ilovezfs commented:
:shipit:

MikeMcQuaid commented:
Will get final 👍 from @UniqMartin and then 🚢

UniqMartin commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -309,6 +314,11 @@ EOS
+  if [[ -z "$HOMEBREW_AUTO_UPDATE_SECS" ]]
+  then
+    HOMEBREW_AUTO_UPDATE_SECS="60"
+  fi
```
If you don't consider this too much voodoo witchcraft, the canonical (and much shorter) way to initialize variables with default values in Bash is to use expansion like this:

```shell
: "${HOMEBREW_AUTO_UPDATE_SECS:=60}"
```

MikeMcQuaid replied:
Will leave it as-is for now just because I think it's a bit more explicit. 👍 to maybe fixing such things in bulk in future though.

UniqMartin commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -378,19 +381,34 @@ EOS
-        # Skip taps without formulae.
-        FORMULAE="$(find "$DIR" -maxdepth 1 \( -name "*.rb" -or -name Formula -or -name HomebrewFormula \) -print -quit)"
-        [[ -z "$FORMULAE" ]] && exit
+        [[ -n "$(find "$DIR/.git/FETCH_HEAD" -type f -mtime -${HOMEBREW_AUTO_UPDATE_SECS}s 2>/dev/null)" ]] && exit
```
`-${HOMEBREW_AUTO_UPDATE_SECS}s` should be quoted for safety and consistency:

```shell
[[ -n "$(find "$DIR/.git/FETCH_HEAD" -type f -mtime "-${HOMEBREW_AUTO_UPDATE_SECS}s" 2>/dev/null)" ]] && exit
```

UniqMartin commented on Library/Homebrew/dev-cmd/test-bot.rb:
```diff
@@ -920,7 +920,6 @@ def sanitize_ARGV_and_ENV
     end
 
     ENV["HOMEBREW_DEVELOPER"] = "1"
-    ENV["HOMEBREW_NO_AUTO_UPDATE"] = "1"
```
Why remove this again? The reason for this addition wasn't speed issues and unless I'm missing something, this PR is only about improving the speed of the auto-update (or update in general).

ilovezfs replied:
"These times are now fast enough to avoid any further special-casing for `--preinstall`"  in commit message

ilovezfs replied:
So I think the idea is subject bot to the same new behavior as users get?

MikeMcQuaid replied:
See https://github.com/Homebrew/brew/pull/669/commits/3c901a972fc57d94763c86965f23e9796a9504c5. It's moved to `brew.sh` instead which means it doesn't need duplicated around every command that needs it.

UniqMartin replied:
I'm confused. How does removing this from the `brew test-bot` code ensure the auto-update doesn't later mess with the state of the repositories once `brew test-bot` has pulled some PR commits or similar activities that involve temporarily modifying the local repository and creating a divergent `master` branch?

From a typical run of `brew test-bot`:

```
# This modifies the master branch locally, making it diverge from upstream:
==> brew pull --clean https://github.com/Homebrew/homebrew-core/pull/3704
[…]
# This causes the auto-update to kick in if it hasn't been disabled in test-bot:
==> brew install --only-dependencies --build-bottle --verbose mesos
```

MikeMcQuaid replied:
> brew install --only-dependencies --build-bottle --verbose mesos

Because the `test-bot` was run through `brew.sh` which has already exported `HOMEBREW_NO_AUTO_UPDATE`. It'll basically handle all cases where a `brew` invocation ends up calling other processes. This also handles the case for e.g. `brew bundle` and other external commands that may call `brew install` in quick succession.

MikeMcQuaid replied:
Basically: you run `brew`: you get one chance for an auto-update within that process or all child processes.

ilovezfs replied:
bin/brew
28:source "$HOMEBREW_LIBRARY/Homebrew/brew.sh"

Q.E.D.

UniqMartin replied:
Thanks for bearing with me! :bow: I misread the code and thought `HOMEBREW_NO_AUTO_UPDATE` would only be exported if a `brew install` or `brew upgrade` command runs, but I see now that it's always set. Sorry for the noise!

UniqMartin commented:
> I've addressed your commits and made it faster still for the `--preinstall` case as a result.

Nice work! I left a few relatively minor comments, but I really like how this has evolved.

(I have a few ideas for how some things could be implemented more neatly, but that's stuff for a follow-up PR rather than a reason to block this. Hopefully I'll find the time to work on this.)

UniqMartin commented:
:+1:

MikeMcQuaid commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -378,19 +381,34 @@ EOS
+    cd "$DIR" || continue
+
+    TAP_VAR="$(repo_var "$DIR")"
+    declare PREFETCH_REVISION"$TAP_VAR"="$(read_current_revision)"
```
This line broke `brew update` for anyone who updated between this commit being merged and the fix in 985c672bac4dc20d369b451c484eb6553762dbcf. It read the current revision of that branch rather than the e.g. `origin/master` revision.

MikeMcQuaid replied:
Open to thoughts on best course of action. My thinking is: mailing list post, Twitter post, IRC room post.

MikeMcQuaid replied:
Did those three. Twitter post is getting decent number of RTs.

ilovezfs replied:
Do the analytics show how many `brew update`s occurred in that period?

UniqMartin replied:
Apologies for the obviously too hasty :+1: as this was a bug I should have been able to spot, but I got too caught up in some minor details and the `test-bot` discussion. 😭

Given how important a functioning `brew update` is for the rest of Homebrew, I think the morale of the story and how we approach future changes to the update should be:

- Don't rush `brew update` changes that aren't urgent bug fixes. (This was about 6 hours from an almost complete rewrite to merging this. Probably too quick.)
- Self-review your own changes after not looking at them for 24 hours.
- If possible, give more co-maintainers a chance to review.
- Do more local testing. If non-obvious, provide instructions how to test this locally, including what you think should happen in various scenarios.

I'm not blaming anyone here. Mistakes happen, but given this incident let's try to learn from them. `brew update` is way too critical to treat changes to it like any other PR.

ilovezfs replied:
The scariest scenario is if this type of bug ever gets pushed out via the auto-update.

ilovezfs replied:
@UniqMartin do you have any idea why `update-test` didn't catch this?

MikeMcQuaid replied:
@ilovezfs I also assumed it would catch this but it doesn't verify the update actually happened until my change in https://github.com/Homebrew/brew/commit/b9d5b11f321a359b462aa2c353c2bdb483930fb3.

MikeMcQuaid replied:
@UniqMartin Agreed with all your learning points. Post 1.0.0 I'm hoping we can slow Homebrew/brew changes right down.

MikeMcQuaid replied:
(and I don't see us needing to change `update` similarly again).

ilovezfs replied:
Yeah, I'm looking at that commit right now. Still surprising to me that it didn't catch it without that commit. My apologies for the :+1: as well :(

MikeMcQuaid replied:
I was also surprised but basically without that commit: there's no error, it just thinks it's up-to-date.

UniqMartin replied:
After a night's sleep and reviewing the changes, I think we might have to make some more tweaks to the logic changed here. 😢 The fix isn't actually fixing the problem, though it's improving the situation to a point where people won't be indefinitely stuck:

```console
$ cd "$(brew --repository)"
$ git checkout master
$ git fetch origin
$ git reset --hard 985c672b # move back to fix commit (not up-to-date with upstream)
HEAD is now at 985c672b update.sh: check upstream SHA prefetch not local.
$ git push --force . 985c672b:origin/master # pretend we didn't fetch the latest commit
Total 0 (delta 0), reused 0 (delta 0)
To .
 + b9d5b11f...985c672b 985c672b -> origin/master (forced update)
$ brew update
Already up-to-date.
$ git rev-parse master
985c672bac4dc20d369b451c484eb6553762dbcf
$ git rev-parse origin/master
b9d5b11f321a359b462aa2c353c2bdb483930fb3
```

Note how `brew update` is claiming up-to-date, but `master` hasn't been updated to `origin/master`. A second `brew update` will actually update, but with the current fix every user will have to run `brew update` twice in a row to actually get updated.

It might also be doing unintended things if something different from `master` is checked out, as it's comparing whatever is currently checked out to `origin/master` (but before any new commits have been fetched and the `origin/master` ref updated).

MikeMcQuaid replied:
@UniqMartin sorted in https://github.com/Homebrew/brew/commit/24b5db53672843647da29073060bff3f81b05b90.
